Cray-HPE/hms-reds
# Crush reds

### Summary and Scope
Major refactoring in REDS:

- removed need for ETCD
- deleted columbia code, it was all broken (because there is no IPAddress data in SLS, and because it was trying to set global IPMI pwds)
- deleted all the SNMP and HTTP handlers
- simplified the HTTP stack

### Issues and Related PRs


* Resolves CASMHMS-5872
* Resolves CASMHMS-4413


## Testing


Tested on:

  * `Mug`
  * Local development environment
  * Virtual Shasta

Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

- Were the install/upgrade-based validation checks/tests run (goss tests/install-validation doc)? Yes
- Were continuous integration tests run? If not, why? Yes
- Was upgrade tested? If not, why? Yes, it works.
- Was downgrade tested? If not, why? Yes, downgrades to the 2.X series from the 3.0 series of charts do not work.  

The following test cases were performed:
- delete master 1 component from HSM and see that it come back after REDS restart
- delete redfish endpoint for some other management NCN and see that it comes back after REDS restart
- for a leaf switch: go into vault, remove leaf switch creds, then validate that they get added back
purge a rosetta from HSM and validate that it gets rediscovered, (by the discovery job).

All test cases were successful. Here are logs and type script from testing:
 
Type script: [reds_testing.log](https://github.com/Cray-HPE/hms-reds-charts/files/10376866/reds_testing.log)
Pod logs:
- [reds_pod.1.log](https://github.com/Cray-HPE/hms-reds-charts/files/10376873/reds_pod.1.log)
- [reds_pod.2.log](https://github.com/Cray-HPE/hms-reds-charts/files/10376874/reds_pod.2.log)
- [reds_pod.3.log](https://github.com/Cray-HPE/hms-reds-charts/files/10376875/reds_pod.3.log)


## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_
 Yes, downgrades to the 2.X series from the 3.0 series of charts do not work.  If a downgrade is required the cray-hms-reds chart needs to be uninstalled, and the 2.X chart needs to be installed. However, the upgrade from 2.Y.Z to 3.Y.Z runs without issues.
```
root@ncn-m001 2023-01-09 19:56:34 ~/anieuwsma # helm -n services rollback cray-hms-reds 1
Error: cannot patch "cray-reds" with kind Deployment: The order in patch list:
[map[name:ETCD_HOST value:cray-reds-etcd-client] map[name:ETCD_HOST value:cray-reds-etcd-client] map[name:ETCD_PORT value:2379] map[name:ETCD_PORT value:2379] map[name:DATASTORE_URL value:$(ETCD_HOST):$(ETCD_PORT)] map[name:REDS_BYPASS_SWITCH_BLACKLIST_PANIC value:] map[name:REDS_SYSLOG_TARG value:rsyslog-aggregator.hmnlb:514] map[name:REDS_NTP_TARG valueFrom:map[configMapKeyRef:map[key:NTP_HOST name:meds-nwproto-info]]]]
 doesn't match $setElementOrder list:
[map[name:ETCD_HOST] map[name:ETCD_PORT] map[name:ETCD_HOST] map[name:ETCD_PORT] map[name:DATASTORE_URL] map[name:HSM_URL] map[name:REDS_BYPASS_SWITCH_BLACKLIST_PANIC] map[name:VAULT_ADDR] map[name:VAULT_SKIP_VERIFY] map[name:SLS_ADDR] map[name:REDS_SYSLOG_TARG] map[name:REDS_NTP_TARG] map[name:REDS_CA_URI]]
```
